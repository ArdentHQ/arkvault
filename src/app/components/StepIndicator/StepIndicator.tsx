import { last } from "@ardenthq/sdk-helpers";
import React, { useMemo } from "react";
import tw, { css, styled } from "twin.macro";

interface StepIndicatorProperties {
	activeIndex?: number;
	steps: string[];
	activeStepTitle?: string;
}

const StepStyled = styled.li<{ isActive: boolean }>`
	${tw`h-0.5 flex-1 rounded-lg transition-colors duration-300`}
	${({ isActive }) =>
		isActive
			? css`
					${tw`bg-theme-warning-300`}
				`
			: css`
					${tw`bg-theme-primary-100 dark:bg-theme-secondary-800`}
				`}
`;

const StepWrapper = styled.ul`
	${tw`flex space-x-3`}
`;

const StepTitle = styled.span`
	${tw`inline-block sm:hidden mx-auto text-theme-secondary-text font-semibold mb-2`}
`;

export const StepIndicator: React.FC<StepIndicatorProperties> = ({
	activeIndex = 1,
	steps,
	activeStepTitle,
}: StepIndicatorProperties) => {
	const title = useMemo(() => {
		if (activeStepTitle) {
			return activeStepTitle;
		}

		if (activeIndex > steps?.length) {
			return last(steps);
		}

		return steps[activeIndex - 1];
	}, [activeIndex, last, steps, activeStepTitle]);

	if (steps.length === 0) {
		return <></>;
	}

	return (
		<div className="flex flex-col">
			<StepTitle>{title}</StepTitle>
			<StepWrapper>
				{steps.map((_, index) => (
					<StepStyled key={index} isActive={activeIndex >= index + 1} />
				))}
			</StepWrapper>
		</div>
	);
};
