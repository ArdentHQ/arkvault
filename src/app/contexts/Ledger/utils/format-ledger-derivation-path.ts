import { LedgerDerivationScheme } from "@/app/contexts/Ledger/Ledger.contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
