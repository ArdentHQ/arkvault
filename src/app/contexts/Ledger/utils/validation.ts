import semver from "semver";
import { Coins } from "@payvo/sdk";
import { minVersionList } from "@/app/contexts/Ledger/Ledger.contracts";

export const hasRequiredAppVersion = async (coin: Coins.Coin) => {
	const coinId = coin.network().coin();

	if (minVersionList[coinId]) {
		const currentVersion = await coin.ledger().getVersion();

		if (semver.lt(currentVersion, minVersionList[coinId])) {
			return false;
		}
	}

	return true;
};
