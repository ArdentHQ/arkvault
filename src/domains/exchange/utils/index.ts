export * from "./assertions";
export * from "./errors";
