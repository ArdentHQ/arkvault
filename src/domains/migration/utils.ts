import { MigrationTransactionStatus } from "@/domains/migration/migration.contracts";

export const getIcon = (status: MigrationTransactionStatus) => {
	if (status === MigrationTransactionStatus.Confirmed) {
		return {
			color: "text-theme-success-600",
			name: "CircleCheckMark",
		};
	}

	return {
		color: "text-theme-warning-300",
		name: "Clock",
	};
};
