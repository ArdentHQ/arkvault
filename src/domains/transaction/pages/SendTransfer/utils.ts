export const getFeeType = (numberOfRecipients: number): "multiPayment" | "transfer" =>
	numberOfRecipients > 1 ? "multiPayment" : "transfer";
