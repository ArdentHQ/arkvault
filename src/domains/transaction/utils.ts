import { Services } from "@ardenthq/sdk";
import { RecipientItem } from "@/domains/transaction/components/RecipientList/RecipientList.contracts";

export const isNoDeviceError = (error: any) => {
	if (!error) {
		return false;
	}

	if (String(error).includes("no device found")) {
		return true;
	}

	return false;
};

export const isRejectionError = (error: any) => String(error).includes("Condition of use not satisfied");

export const handleBroadcastError = ({ errors }: Services.BroadcastResponse) => {
	const allErrors = Object.values(errors);

	if (allErrors.length === 0) {
		return;
	}

	throw new Error(allErrors[0]);
};

export const getTransferType = ({ recipients }: { recipients: RecipientItem[] }): "multiPayment" | "transfer" => {
	const isMultiPayment = recipients.length > 1;
	return isMultiPayment ? "multiPayment" : "transfer";
};

export const withAbortPromise =
	(signal?: AbortSignal, callback?: () => void) =>
	<T>(promise: Promise<T>) =>
		new Promise<T>((resolve, reject) => {
			if (signal) {
				signal.addEventListener("abort", () => {
					callback?.();
					reject("ERR_ABORT");
				});
			}

			return promise.then(resolve).catch(reject);
		});
