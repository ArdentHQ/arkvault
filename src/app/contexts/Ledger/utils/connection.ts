import { Coins } from "@ardenthq/sdk";
import retry, { AbortError, Options } from "p-retry";

import { formatLedgerDerivationPath } from "./format-ledger-derivation-path";
import { hasRequiredAppVersion } from "./validation";

export const accessLedgerDevice = async (coin: Coins.Coin) => {
	try {
		await coin.__construct();
		await coin.ledger().connect();
	} catch (error) {
		// If the device is open, continue normally.
		// Can be triggered when the user retries ledger connection.
		if (error.message !== "The device is already open.") {
			throw error;
		}
	}
};

export const accessLedgerApp = async ({ coin }: { coin: Coins.Coin }) => {
	await accessLedgerDevice(coin);

	if (!(await hasRequiredAppVersion(coin))) {
		throw new Error("VERSION_ERROR");
	}

	// Ensure that the app is accessible.
	await coin.ledger().getPublicKey(
		formatLedgerDerivationPath({
			coinType: coin.config().get<number>("network.constants.slip44"),
		}),
	);
};

export const persistLedgerConnection = async ({
	coin,
	options,
	hasRequestedAbort,
}: {
	coin: Coins.Coin;
	options: Options;
	hasRequestedAbort: () => boolean;
}) => {
	const retryAccess: any = async (attempts: number) => {
		if (hasRequestedAbort() && attempts > 1) {
			throw new AbortError("CONNECTION_ERROR");
		}

		try {
			await accessLedgerApp({ coin });
		} catch (error) {
			// Abort on version error or continue retrying access.
			if (error.message === "VERSION_ERROR") {
				throw new AbortError("VERSION_ERROR");
			}

			throw error;
		}
	};

	await retry(retryAccess, options);
};
