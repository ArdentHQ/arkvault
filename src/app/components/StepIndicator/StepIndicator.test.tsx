import React from "react";
import { StepIndicator } from "./StepIndicator";
import { render, screen } from "@/utils/testing-library";

describe("StepIndicator", () => {
	it("should render", () => {
		const { asFragment } = render(<StepIndicator steps={["1", "2", "3"]} />);

		expect(screen.getByRole("list")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render witth active index defined", () => {
		const { asFragment } = render(<StepIndicator steps={["1", "2", "3"]} activeIndex={2} />);

		expect(screen.getByRole("list")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should limit active index to max number of steps", () => {
		const { asFragment } = render(<StepIndicator steps={["1", "2", "3"]} activeIndex={5} />);

		expect(screen.getByRole("list")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should not render with an empty steps array", () => {
		const { asFragment } = render(<StepIndicator steps={[]} />);

		expect(screen.queryByRole("list")).not.toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});
});
