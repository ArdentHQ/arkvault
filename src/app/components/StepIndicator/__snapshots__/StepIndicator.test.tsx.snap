// Vitest Snapshot v1

exports[`StepIndicator > should limit active index to max number of steps 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
  >
    <span
      class="css-3qwsx5"
    >
      3
    </span>
    <ul
      class="css-xa5o6q"
    >
      <li
        class="css-47028"
      />
      <li
        class="css-47028"
      />
      <li
        class="css-47028"
      />
    </ul>
  </div>
</DocumentFragment>
`;

exports[`StepIndicator > should not render with an empty steps array 1`] = `<DocumentFragment />`;

exports[`StepIndicator > should render 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
  >
    <span
      class="css-3qwsx5"
    >
      1
    </span>
    <ul
      class="css-xa5o6q"
    >
      <li
        class="css-47028"
      />
      <li
        class="css-3ldhk3"
      />
      <li
        class="css-3ldhk3"
      />
    </ul>
  </div>
</DocumentFragment>
`;

exports[`StepIndicator > should render witth active index defined 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
  >
    <span
      class="css-3qwsx5"
    >
      2
    </span>
    <ul
      class="css-xa5o6q"
    >
      <li
        class="css-47028"
      />
      <li
        class="css-47028"
      />
      <li
        class="css-3ldhk3"
      />
    </ul>
  </div>
</DocumentFragment>
`;
